src/symbol/place_symbols.js

```diff
@@ -0,0 +1,260 @@
+// @flow
+
+const symbolSize = require('./symbol_size');
+
+import type SymbolBucket, {SymbolInstance} from '../data/bucket/symbol_bucket';
+import type OpacityState from './opacity_state';
+import type CollisionIndex from './collision_index';
+import type CollisionBoxArray from './collision_box';
+import type {StructArray} from '../util/struct_array';
+const mat4 = require('@mapbox/gl-matrix').mat4;
+
+module.exports = {
+    updateOpacities: updateOpacities,
+    place: place
+};
+
+function updateOpacity(symbolInstance: SymbolInstance, opacityState: OpacityState, targetOpacity: number, opacityUpdateTime: number, collisionFadeTimes: any): boolean {
+    const initialVisible = opacityState.opacity !== 0;
+    const initialTargetVisible = opacityState.targetOpacity !== 0;
+    if (symbolInstance.isDuplicate) {
+        opacityState.opacity = 0;
+        opacityState.targetOpacity = 0;
+    } else {
+        if (opacityState.targetOpacity !== targetOpacity) {
+            collisionFadeTimes.latestStart = opacityUpdateTime;
+        }
+        const increment = collisionFadeTimes.duration ? ((opacityUpdateTime - opacityState.time) / collisionFadeTimes.duration) : 1;
+        opacityState.opacity = Math.max(0, Math.min(1, opacityState.opacity + (opacityState.targetOpacity === 1 ? increment : -increment)));
+        opacityState.targetOpacity = targetOpacity;
+        opacityState.time = opacityUpdateTime;
+    }
+    return (!initialVisible && !initialTargetVisible && opacityState.targetOpacity !== 0) ||
+           ((initialVisible || initialTargetVisible) && (opacityState.opacity === 0 && opacityState.targetOpacity === 0));
```

Comment:
I think this is *right* but it could be clearer as
```js
const initialVisible = opacityState.opacity !== 0 && opacityState.targetOpacity !== 0;
const nowVisible = opacityState.opacity !== 0 && opacityState.targetOpacity !== 0;
return initialVisible !== nowVisible
```

also, the return value is only used for text. Should it be moved out of this function and just around the text?

Reply:
"visible" would be `opacity != 0 || targetOpacity != 0`, but I changed it instead to just `initialHidden` and `nowHidden`, and moved them out to the test label call as you suggested.